packaging/_parser.py

```diff
@@ -0,0 +1,199 @@
+from ast import literal_eval
+from typing import Any, List, NamedTuple, Tuple, Union
+
+from ._tokenizer import Tokenizer
+
+
+class Node:
+    def __init__(self, value: str) -> None:
+        self.value = value
+
+    def __str__(self) -> str:
+        return self.value
+
+    def __repr__(self) -> str:
+        return f"<{self.__class__.__name__}('{self}')>"
+
+    def serialize(self) -> str:
+        raise NotImplementedError
+
+
+class Variable(Node):
+    def serialize(self) -> str:
+        return str(self)
+
+
+class Value(Node):
+    def serialize(self) -> str:
+        return f'"{self}"'
+
+
+class Op(Node):
+    def serialize(self) -> str:
+        return str(self)
+
+
+MarkerVar = Union[Variable, Value]
+MarkerItem = Tuple[MarkerVar, Op, MarkerVar]
+# MarkerAtom = Union[MarkerItem, List["MarkerAtom"]]
+# MarkerList = List[Union["MarkerList", MarkerAtom, str]]
+# mypy does not suport recursive type definition
+# https://github.com/python/mypy/issues/731
+MarkerAtom = Any
+MarkerList = List[Any]
+
+
+class Requirement(NamedTuple):
+    name: str
+    url: str
+    extras: List[str]
+    specifier: str
+    marker: str
+
+
+def parse_named_requirement(requirement: str) -> Requirement:
+    """
+    NAMED_REQUIREMENT: NAME EXTRAS* URL_SPEC (SEMICOLON + MARKER_EXPR)*
+    """
+    tokens = Tokenizer(requirement)
+    tokens.expect("IDENTIFIER", error_message="Expression must begin with package name")
+    name = tokens.read("IDENTIFIER").text
+    extras = parse_extras(tokens)
+    specifier = ""
+    url = ""
+    if tokens.match("URL_SPEC"):
+        url = tokens.read().text[1:].strip()
+    elif not tokens.match("stringEnd"):
+        specifier = parse_specifier(tokens)
+    if tokens.try_read("SEMICOLON"):
+        marker = ""
+        while not tokens.match("stringEnd"):
+            # we don't validate markers here, it's done later as part of
+            # packaging/requirements.py
+            marker += tokens.read().text
+    else:
+        marker = ""
+        tokens.expect("stringEnd", error_message="Expected string_end or semicolon")
+    return Requirement(name, url, extras, specifier, marker)
+
+
+def parse_extras(tokens: Tokenizer) -> List[str]:
+    """
+    EXTRAS: (LBRACKET + IDENTIFIER + (COMMA + IDENTIFIER)* + RBRACKET)*
+    """
+    extras = []
+    if tokens.try_read("LBRACKET"):
+        while tokens.match("IDENTIFIER"):
+            extras.append(tokens.read("IDENTIFIER").text)
+            if not tokens.match("RBRACKET"):
+                tokens.expect("COMMA", error_message="Missing comma after extra")
+                tokens.read("COMMA")
+            if not tokens.match("COMMA") and tokens.match("RBRACKET"):
```

Comment:
```suggestion
            if tokens.match("RBRACKET"):
```

We don't want to allow `,,,` in extras? This should get a corresponding test too, IMO.

Reply:
Ah, nvm. I misread this. :)